src/appleseed/foundation/math/dual.h

```diff
@@ -102,6 +104,13 @@ inline const T& Dual<T>::get_value() const
 }
 
 template <typename T>
+inline void Dual<T>::set_value(const T& value)
+{
+    m_value = value;
+    m_has_derivatives = false;
```

Comment:
That's dangerous, e.g. if `set_value()` is called after `set_derivatives()`. I would recommend against clearing the flag here.